### ISSUE TITLE
Intelligent switching between 1080p and 4K

### ISSUE THREAD

**Issue by diamondsw** [JohnCoates/Aerial#503]
When I first installed the beta of Aerial for Mojave and tvOS 12 screen saver support, I was greeted with slideshow performance on my 3-display 2013 MacBook Pro. This all cleared up when I switched from the new 4K default back to 1080p - my poor MacBook with integrated graphics can't push that much video apparently. However, the only reason I knew that toggle was there was because I'd been reading a few issue threads in here. I imagine a lot of people won't know this, and the default to 4K will not serve them well. All they know is performance tanked.

I'd request an "Auto" setting to dynamically choose 1080p videos for any screen 1920px wide (or less), and 4K videos for screens above that resolution. No sense in pushing 4K video where it's not visible. If possible, handle this on a monitor-by-monitor basis.

**Comment by glouel** [comment]
The first beta only had 4K support, then I added back 1080p. Because of that I kept it as default.

However I definitely agree, 1080p should probably even be the default for everyone (the whole 4K set takes 12.5 GB which may be an issue for a lot of people by itself).  

Side note, the new videos (1080p and 4K) are all using HEVC instead of H264. H.264 is still available for 1080p but I haven't added the option to use them yet. HEVC has a pretty big size advantage. On older configurations, HEVC may be an issue. In your specific case, I guess you have a dedicated GPU on your MBP (that may handle 1080p HEVC but not 4K or not 3x 4K streams). Haswell's iGPU doesn't hw decode HEVC at all, so that's probably an issue that needs fixing too for those machines and older.

**Comment by eivindml** [comment]
Is it only 1080p at this point? It looks really grainy on my LG 5K display. I wouldn't have a problem with using 12GB.

**Comment by glouel** [comment]
If you are using my builds from the other thread, check in the cache panel if you are using 4K. Some videos are grainy/blurry (especially space), some aren't. But if it's 4K, it should only use 4K. If you want to compare, you can go manually in the cache folder and check the videos. All the new ones you have downloaded should be ending in something like "HEVC_4K.mov" (you may also have some previous videos in that folder but they won't be used). 1080p shows as "HEVC 2K".

**Comment by diamondsw** [comment]
The 13" MacBook Pro only had integrated graphics, and being as old as it is I'm pretty sure I don't have HEVC capability. Screen savers display fine on 1080p for me, including the lovely new ISS ones.

**Comment by diamondsw** [comment]
I should also note, checking the horizontal resolution is probably a better idea than vertical, owing to so many 16x10 monitors out there. Otherwise you'll scale up to 4K needlessly on the common 1920x1200 resolution.

**Comment by glouel** [comment]
Well, HEVC should still play on your Mac but it will be CPU decoded instead of GPU accelerated, so you may see high CPU load (see this issue https://github.com/JohnCoates/Aerial/issues/505 ) 

I've added a build here with 1080p H264 as default : 
https://github.com/JohnCoates/Aerial/issues/463#issuecomment-424407877

We'll see what we can do about detection at a later point.

**Comment by BreiteSeite** [comment]
I also have moderate cpu and high gpu load on my 2016 MacBook TouchBar with the Radeon Pro 455 connected to the LG UltraFine 5k display on macOS 10.14. It ramps up the fan very noticeably. It sometimes even drops frame.

I cached all videos in 4k HEVC.

Shouldn't it be very smooth and non-taxin, due to hardware HEVC decoding support?

**Comment by glouel** [comment]
@BreiteSeite I tried to investigate that a bit in that issue https://github.com/JohnCoates/Aerial/issues/512#issuecomment-426052836

In your case, Late 2016, that's still a Skylake CPU so you only have "semi" accelerated HEVC decoding, because the videos are encoded in Main10 HEVC profile. 

As for your Radeon Pro 455, in theory it should have HEVC 10 bit playback, whether it is used or not, I don't know. It may not be used at all, or because you have two gpus, only some of the time. I'd suggest as I did in the linked comment checking CPU usage when playing the same video on QuickTime, see what happens ?

**Comment by BreiteSeite** [comment]
@glouel 1st: thank you for the awesome and really quick issue support/reaction!

I already tried playing the video via quicktime. And in activity monitor, you can really see how taxing it is to the GPU.

<img width="359" alt="screenshot 2018-10-02 at 19 48 54" src="https://user-images.githubusercontent.com/2171105/46366753-3c9d8280-c67c-11e8-8da2-17abc77b6381.png">

This is with the 5k display connected, which forces the discrete gpu anyway.

**Comment by glouel** [comment]
@BreiteSeite I'm not sure that 100% means it uses the GPU for decoding. The decoding is based on fixed function units, not on the GPU itself so it's hard to say what that GPU load is (I'm not sure what sensor is used). Note that the GPU itself doesn't have to be active in order for Intel's QuickSync to be used so forcing the discrete GPU may not affect MacOS's GVA video decoding. 

In any case, you still have a fair bit of CPU load. I'd suggest you check with Activity Monitor, see what CPU load you have on VTDecoderXPCService with QuickTime, that would give us a good clue.

**Comment by BreiteSeite** [comment]
@glouel <img width="815" alt="screenshot 2018-10-02 at 20 00 46" src="https://user-images.githubusercontent.com/2171105/46367446-dfa2cc00-c67d-11e8-9c5b-e9f10d2adead.png">

**Comment by glouel** [comment]
@BreiteSeite Thanks !

So I'd say it's definitely not fully GPU accelerated, and probably not playing through your Polaris UVD. I guess it uses Intel's QuickSync by default, your CPU load looks close to what I have on my iMac which also lacks Main10 decoding on QuickSync (and on my GPU). 

I did check the GPU load thing, and as I suspected, you see a "GPU load" just from the video playback itself (I guess from the buffers). Here's what I have with QuickSync "assisted" (but not complete) HEVC decoding : 

<img width="1129" alt="capture d ecran 2018-10-02 a 20 01 38" src="https://user-images.githubusercontent.com/37544189/46367667-812a1d80-c67e-11e8-805e-d2f42adc3895.png">

If you try a 1080p H.264 video you'll probably see VTDecoderXPCService below 5%.
